interfaces/apparmor/spec.go

```diff
@@ -173,6 +174,92 @@ func isProbablyPresent(path string) bool {
+// expression describes all the files and directories at that level. Coupled
+// with the string "rw ," it can be used to create a rule that allows write
+// access to any file therein, but not deeper.
+//
+// For example, with path: "/foo/bar/froz/baz" and depth 3 the result is:
+// []string{"/", "/foo/", "/foo/bar/"} and []string{"/foo/*", "/foo/bar/*",
+// "/foo/bar/froz/*"}. Coupled with the aforementioned constants this
+// translates to the following apparmor rules:
+//
+//   / r,
+//   /foo/ r,
+//   /foo/bar/ r,
+//
+//   /foo/bar/* rw,
+//   /foo/bar/*/ rw,
+//   /foo/bar/froz/* rw,
+//   /foo/bar/froz/*/ rw,
+//
+// Those rules are useful for constructing the apparmor profile for a writable
+// mimic that needs to be present in a specific directory (e.g. in
+// /foo/bar/froz/baz) assuming that part of that directory already exists (e.g.
+// /foo/bar/) but may need to be created earlier (e.g. in /foo/bar/froz).
+//
+// The mimic works by mounting a tmpfs over the mimicked directory and then
+// re-creating empty files and directories as mount points for the subsequent
+// bind- mount operations to latch onto. This is why the right list of
+// expressions use * and */, this allows the expressions to capture files and
+// directories at a specific path.
+func chopTree(path string, assumedPrefixDepth int) (left, right []string, err error) {
+	// NOTE: This implementation works around a bug in apparmor parser:
+	// https://bugs.launchpad.net/apparmor/+bug/1769971
+	//
+	// Due to the nature of apparmor path expressions we need to distinguish
+	// directories and files. The path expression denoting a directory must end
+	// with a trailing slash, that denoting a file must not.
+	//
+	// The iterator requires golang-clean paths which never have a trailing
+	// slash. We want to allow clean paths with an optional trailing slash.
+	isDir := strings.HasSuffix(path, "/")
+	cleanPath := filepath.Clean(path)
+	if (isDir && cleanPath+"/" != path) || (!isDir && cleanPath != path) {
+		return nil, nil, fmt.Errorf("cannot chop unclean path: %q", path)
+	}
+
+	// Iterate over the path and construct left and right.
+	iter, _ := strutil.NewPathIterator(cleanPath)
+	for iter.Next() {
+		if iter.Depth() <= assumedPrefixDepth {
+			// The left hand side is the part that is assumed to exist.
+			// We mostly enumerate those directories as-is except for the final
+			// entry that we re-create the trailing slash if the original path
+			// was a "directory" path.
+			if iter.CurrentPath() == iter.Path() && isDir {
+				left = append(left, iter.CurrentPath()+"/")
+			} else {
+				left = append(left, iter.CurrentPath())
+			}
+		} else {
+			// The right hand side replaces the final component with a "*" and
+			// "*/", meaning any file and any directory, respectively.
+			if iter.Depth() > 1 {
+				// The right hand side rules should not allow creation of the
+				// root directory as that itself is meaningless.
```

Comment:
This comment should be moved up as it is talking about '> 1' and not the append operations.